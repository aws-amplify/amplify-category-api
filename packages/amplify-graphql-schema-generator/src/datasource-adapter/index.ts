export * from './datasource-adapter';
export * from './string-datasource-adapter';
export { MySQLDataSourceAdapter } from './mysql-datasource-adapter';
export { PostgresDataSourceAdapter } from './pg-datasource-adapter';
export { MySQLStringDataSourceAdapter } from './mysql-string-datasource-adapter';
export { PostgresStringDataSourceAdapter } from './pg-string-datasource-adapter';
