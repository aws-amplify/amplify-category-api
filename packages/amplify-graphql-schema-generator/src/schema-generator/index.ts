export * from './generate-schema';
export * from './schema-overrides';
