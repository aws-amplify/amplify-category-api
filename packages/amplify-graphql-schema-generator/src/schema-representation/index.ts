export * from './engine';
export * from './schema';
export * from './types';
