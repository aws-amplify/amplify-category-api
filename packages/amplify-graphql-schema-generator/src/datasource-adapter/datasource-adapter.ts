import * as os from 'os';
import { Field, Index, Model } from '../schema-representation';

export interface DataSourceConfig {
  host: string;
  port: number;
  database: string;
  username: string;
  password: string;
  sslCertificate?: string;
}

export abstract class DataSourceAdapter {
  public abstract getTablesList(): string[];

  public abstract getFields(tableName: string): Field[];

  public abstract getPrimaryKey(tableName: string): Index | null;

  public abstract getIndexes(tableName: string): Index[];

  public abstract initialize(): Promise<void>;

  public abstract cleanup(): void;

  public abstract test(): Promise<boolean>;

  protected abstract querySchema(): Promise<string>;

  public useVPC = false;

  public vpcSchemaInspectorLambda: string | undefined = undefined;

  public vpcLambdaRegion: string | undefined = undefined;

  public getModels(): Model[] {
    const tableNames = this.getTablesList();
    const models = [];
    for (const table of tableNames) {
      models.push(this.describeTable(table));
    }
    return models;
  }

  public describeTable(tableName: string): Model {
    // Retrieve the fields, primary key and indexes info
    const fields = this.getFields(tableName);
    const primaryKey = this.getPrimaryKey(tableName);
    const indexes = this.getIndexes(tableName);

    // Construct the model from the retrieved details
    const model = new Model(tableName);
    fields.forEach((field) => model.addField(field));
    primaryKey && model.setPrimaryKey(primaryKey.getFields());
    indexes.forEach((index) => model.addIndex(index.name, index.getFields()));
    return model;
  }

  public useVpc(vpcSchemaInspectorLambda: string, region: string): void {
    this.useVPC = true;
    this.vpcSchemaInspectorLambda = vpcSchemaInspectorLambda;
    this.vpcLambdaRegion = region;
  }

  protected queryToCSV(queryResult: any[]): string {
    if (queryResult.length === 0) {
      return '';
    }
    const headers = Object.keys(queryResult[0]);
    const headerIndices = Object.fromEntries(headers.map((key, index) => [index, key]));
    const rows = queryResult.map((row) =>
      [...Array(headers.length).keys()]
        .map((index) => {
          const value = row[headerIndices[index]];
          // sanitize if comma is present in value
          if (typeof value === 'string' && value.includes(',')) {
            return `"${value}"`;
          }
          if (value === null) {
            return 'NULL';
          }
          return value;
        })
        .join(','),
    );
    return headers.join(',') + os.EOL + rows.join(os.EOL);
  }
}
