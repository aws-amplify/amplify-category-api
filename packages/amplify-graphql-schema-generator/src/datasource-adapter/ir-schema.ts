import { Field, Index } from '../schema-representation';

export type IRSchemaInputs = {
  fields: any;
  indexes: any;
  tables: any;
};

export abstract class IRSchema {
  protected abstract setFields(fields: string): void;
  protected abstract setIndexes(indexes: string): void;
  protected abstract setTables(tables: string): void;
  public abstract getFields(tableName: string): Field[];
  public abstract getPrimaryKey(tableName: string): Index | null;
  public abstract getIndexes(tableName: string): Index[];

  constructor({ fields, indexes, tables }: IRSchemaInputs) {
    this.setFields(fields);
    this.setIndexes(indexes);
    this.setTables(tables);
  }
}
