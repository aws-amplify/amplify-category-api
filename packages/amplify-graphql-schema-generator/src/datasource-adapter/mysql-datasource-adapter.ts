import { knex } from 'knex';
import ora from 'ora';
import { invokeSchemaInspectorLambda } from '../utils/vpc-helper';
import { Field, Index } from '../schema-representation';
import { getSSLConfig } from '../utils';
import { DataSourceAdapter } from './datasource-adapter';
import { MySQLStringDataSourceAdapter } from './mysql-string-datasource-adapter';

const spinner = ora();
export interface MySQLDataSourceConfig {
  host: string;
  port: number;
  database: string;
  username: string;
  password: string;
}

export class MySQLDataSourceAdapter extends DataSourceAdapter {
  private adapter: MySQLStringDataSourceAdapter;
  private dbBuilder: any;

  constructor(private config: MySQLDataSourceConfig) {
    super();
  }

  public async test(): Promise<boolean> {
    const TEST_QUERY = 'SELECT 1';
    if (!this.dbBuilder) {
      this.establishDBConnection();
    }
    try {
      await this.dbBuilder.raw(TEST_QUERY);
    } catch (error) {
      return false;
    }
    return true;
  }

  public async initialize(): Promise<void> {
    spinner.start('Fetching the database schema...');
    try {
      await this.establishDBConnection();
      const schema = await this.querySchema();
      this.adapter = new MySQLStringDataSourceAdapter(schema);
    } catch (error) {
      spinner.fail('Failed to fetch the database schema.');
      throw error;
    }
    spinner.succeed('Successfully fetched the database schema.');
  }

  private establishDBConnection(): void {
    const databaseConfig = {
      host: this.config.host,
      database: this.config.database,
      port: this.config.port,
      user: this.config.username,
      password: this.config.password,
      ssl: getSSLConfig(this.config.host),
    };
    try {
      this.dbBuilder = knex({
        client: 'mysql2',
        connection: databaseConfig,
        pool: {
          min: 5,
          max: 30,
          createTimeoutMillis: 30000,
          acquireTimeoutMillis: 30000,
          idleTimeoutMillis: 30000,
          reapIntervalMillis: 1000,
          createRetryIntervalMillis: 100,
        },
        debug: false,
      });
    } catch (err) {
      console.info(err);
      throw err;
    }
  }

  public getTablesList(): string[] {
    return this.adapter.getTablesList();
  }

  public getFields(tableName: string): Field[] {
    return this.adapter.getFields(tableName);
  }

  public getPrimaryKey(tableName: string): Index | null {
    return this.adapter.getPrimaryKey(tableName);
  }

  public getIndexes(tableName: string): Index[] {
    return this.adapter.getIndexes(tableName);
  }

  protected async querySchema(): Promise<string> {
    const schemaQuery = getMySQLSchemaQuery(this.config.database);
    const result =
      this.useVPC && this.vpcSchemaInspectorLambda
        ? await invokeSchemaInspectorLambda(this.vpcSchemaInspectorLambda, this.config, schemaQuery, this.vpcLambdaRegion)
        : (await this.dbBuilder.raw(schemaQuery))[0];
    return this.queryToCSV(result);
  }

  public cleanup(): void {
    this.dbBuilder && this.dbBuilder.destroy();
  }
}

export function getMySQLSchemaQuery(databaseName: string): string {
  return `
SELECT DISTINCT
  INFORMATION_SCHEMA.COLUMNS.TABLE_NAME,
  INFORMATION_SCHEMA.COLUMNS.COLUMN_NAME,
  INFORMATION_SCHEMA.COLUMNS.COLUMN_DEFAULT,
  INFORMATION_SCHEMA.COLUMNS.ORDINAL_POSITION,
  INFORMATION_SCHEMA.COLUMNS.DATA_TYPE,
  INFORMATION_SCHEMA.COLUMNS.COLUMN_TYPE,
  INFORMATION_SCHEMA.COLUMNS.IS_NULLABLE,
  INFORMATION_SCHEMA.COLUMNS.CHARACTER_MAXIMUM_LENGTH,
  INFORMATION_SCHEMA.STATISTICS.INDEX_NAME,
  INFORMATION_SCHEMA.STATISTICS.NON_UNIQUE,
  INFORMATION_SCHEMA.STATISTICS.SEQ_IN_INDEX,
  INFORMATION_SCHEMA.STATISTICS.NULLABLE
FROM INFORMATION_SCHEMA.COLUMNS
LEFT JOIN INFORMATION_SCHEMA.STATISTICS ON INFORMATION_SCHEMA.COLUMNS.TABLE_NAME=INFORMATION_SCHEMA.STATISTICS.TABLE_NAME AND INFORMATION_SCHEMA.COLUMNS.COLUMN_NAME=INFORMATION_SCHEMA.STATISTICS.COLUMN_NAME
WHERE INFORMATION_SCHEMA.COLUMNS.TABLE_SCHEMA = '${databaseName}'
`;
}
