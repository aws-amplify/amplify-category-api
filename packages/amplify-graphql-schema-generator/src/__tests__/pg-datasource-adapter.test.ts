import { DataSourceAdapter, PostgresDataSourceAdapter } from '../datasource-adapter';
import { Engine, Field, FieldType, Index, Model, Schema } from '../schema-representation';
import { generateGraphQLSchema, isComputeExpression } from '../schema-generator';
import { gql } from 'graphql-transformer-core';

class TestDataSourceAdapter extends DataSourceAdapter {
  public async initialize(): Promise<void> {
    // Do Nothing
  }

  public mapDataType(type: string, nullable: boolean): FieldType {
    return {
      kind: 'Scalar',
      name: 'String',
    };
  }

  public getTablesList(): string[] {
    return ['Test'];
  }

  public getFields(tableName: string): Field[] {
    return [];
  }

  public getPrimaryKey(tableName: string): Index | null {
    return null;
  }

  public getIndexes(tableName: string): Index[] {
    return [];
  }

  public cleanup(): void {
    // Do Nothing
  }
  public async test(): Promise<boolean> {
    return true;
  }

  protected async querySchema(): Promise<string> {
    return '';
  }
}

describe('testPostgresDataSourceAdapter', () => {
  it('getModels call the default implementation', async () => {
    const adapter: DataSourceAdapter = new TestDataSourceAdapter();
    adapter.getTablesList = jest.fn(() => ['Test']);
    adapter.getFields = jest.fn(() => []);
    adapter.getPrimaryKey = jest.fn();
    adapter.getIndexes = jest.fn(() => []);
    await adapter.getModels();
    expect(adapter.getTablesList).toBeCalledTimes(1);
    expect(adapter.getFields).toBeCalledTimes(1);
    expect(adapter.getIndexes).toBeCalledTimes(1);
    expect(adapter.getPrimaryKey).toBeCalledTimes(1);
  });

  it('test generate graphql schema from internal reprensentation', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    let model = new Model('Capital');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('countryId', { kind: 'Scalar', name: 'Int' }));
    model.setPrimaryKey(['id']);
    model.addIndex('countryId', ['countryId']);
    dbschema.addModel(model);

    model = new Model('Country');
    const countryIdField = new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    countryIdField.default = { kind: 'DB_GENERATED', value: 'uuid()' };
    model.addField(countryIdField);
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Tasks');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('title', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('description', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('priority', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id', 'title']);
    model.addIndex('tasks_title', ['title']);
    model.addIndex('tasks_title_description', ['title', 'description']);
    dbschema.addModel(model);

    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('include option should import only the given tables', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    let model = new Model('Capital');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('countryId', { kind: 'Scalar', name: 'Int' }));
    model.setPrimaryKey(['id']);
    model.addIndex('countryId', ['countryId']);
    dbschema.addModel(model);

    model = new Model('Country');
    const countryIdField = new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    countryIdField.default = { kind: 'DB_GENERATED', value: 'uuid()' };
    model.addField(countryIdField);
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Tasks');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('title', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('description', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('priority', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id', 'title']);
    model.addIndex('tasks_title', ['title']);
    model.addIndex('tasks_title_description', ['title', 'description']);
    dbschema.addModel(model);

    const amplifyInputType = gql`
      input AMPLIFY {
        engine: String = "postgres"
        globalAuthRule: AuthRule = { allow: public }
        include: [String] = ["Tasks"]
      }
    `;

    const graphqlSchema = generateGraphQLSchema(dbschema, amplifyInputType);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('exclude option should not import the given tables', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    let model = new Model('Capital');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('countryId', { kind: 'Scalar', name: 'Int' }));
    model.setPrimaryKey(['id']);
    model.addIndex('countryId', ['countryId']);
    dbschema.addModel(model);

    model = new Model('Country');
    const countryIdField = new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    countryIdField.default = { kind: 'DB_GENERATED', value: 'uuid()' };
    model.addField(countryIdField);
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Tasks');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('title', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('description', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('priority', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id', 'title']);
    model.addIndex('tasks_title', ['title']);
    model.addIndex('tasks_title_description', ['title', 'description']);
    dbschema.addModel(model);

    const amplifyInputType = gql`
      input AMPLIFY {
        engine: String = "postgres"
        globalAuthRule: AuthRule = { allow: public }
        exclude: [String] = ["Tasks"]
      }
    `;

    const graphqlSchema = generateGraphQLSchema(dbschema, amplifyInputType);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('providing both include and exclude option should throw an error', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    let model = new Model('Capital');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('countryId', { kind: 'Scalar', name: 'Int' }));
    model.setPrimaryKey(['id']);
    model.addIndex('countryId', ['countryId']);
    dbschema.addModel(model);

    model = new Model('Country');
    const countryIdField = new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    countryIdField.default = { kind: 'DB_GENERATED', value: 'uuid()' };
    model.addField(countryIdField);
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Tasks');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('title', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('description', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('priority', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id', 'title']);
    model.addIndex('tasks_title', ['title']);
    model.addIndex('tasks_title_description', ['title', 'description']);
    dbschema.addModel(model);

    const amplifyInputType = gql`
      input AMPLIFY {
        engine: String = "postgres"
        globalAuthRule: AuthRule = { allow: public }
        include: [String] = ["Tasks"]
        exclude: [String] = ["Tasks"]
      }
    `;
    expect(() => generateGraphQLSchema(dbschema, amplifyInputType)).toThrowError(
      'Cannot specify both include and exclude options. Please check your GraphQL schema.',
    );
  });

  it('providing incorrect include and exclude datatype should throw an error', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    let model = new Model('Capital');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('countryId', { kind: 'Scalar', name: 'Int' }));
    model.setPrimaryKey(['id']);
    model.addIndex('countryId', ['countryId']);
    dbschema.addModel(model);

    model = new Model('Country');
    const countryIdField = new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    countryIdField.default = { kind: 'DB_GENERATED', value: 'uuid()' };
    model.addField(countryIdField);
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Tasks');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('title', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('description', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('priority', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id', 'title']);
    model.addIndex('tasks_title', ['title']);
    model.addIndex('tasks_title_description', ['title', 'description']);
    dbschema.addModel(model);

    const amplifyInputTypeInclude = gql`
      input AMPLIFY {
        engine: String = "postgres"
        globalAuthRule: AuthRule = { allow: public }
        include: String = "Tasks"
      }
    `;
    expect(() => generateGraphQLSchema(dbschema, amplifyInputTypeInclude)).toThrowError(
      'Invalid value for include option. Please check your GraphQL schema.',
    );

    const amplifyInputTypeExclude = gql`
      input AMPLIFY {
        engine: String = "postgres"
        globalAuthRule: AuthRule = { allow: public }
        exclude: String = "Tasks"
      }
    `;
    expect(() => generateGraphQLSchema(dbschema, amplifyInputTypeExclude)).toThrowError(
      'Invalid value for include option. Please check your GraphQL schema.',
    );
  });

  it('generate schema retains hasOne and belongsTo relationship and removes the non-relational fields added manually', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    let model = new Model('User');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Profile');
    const profileIdField = new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    model.addField(profileIdField);
    model.addField(new Field('content', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('userId', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const existingSchema = gql`
      input AMPLIFY {
        engine: String = "postgres"
        globalAuthRule: AuthRule = { allow: public }
      }
      type User @model {
        id: Int!
        name: String
        manuallyAddedField: String
        profile: Profile @hasOne(references: ["userId"])
      }
      type Profile @model {
        id: Int!
        content: String
        manuallyAddedField: String
        userId: Int!
        user: User @belongsTo(references: ["userId"])
      }
    `;

    const graphqlSchema = generateGraphQLSchema(dbschema, existingSchema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('generate schema retains hasMany and belongsTo relationship', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    let model = new Model('Blog');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('Post');
    const postIdField = new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    model.addField(postIdField);
    model.addField(new Field('content', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('blogId', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const existingSchema = gql`
      input AMPLIFY {
        engine: String = "postgres"
        globalAuthRule: AuthRule = { allow: public }
      }
      type Blog @model {
        id: Int!
        name: String
        posts: [Post] @hasMany(references: ["userId"])
      }
      type Post @model {
        id: Int!
        content: String
        blogId: Int!
        blog: Blog @belongsTo(references: ["userId"])
      }
    `;

    const graphqlSchema = generateGraphQLSchema(dbschema, existingSchema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('generates a default directive and optional types for fields with literal default values', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    const model = new Model('Account');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    const serialNoField = new Field('serialNumber', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    const ownerNameField = new Field('ownerName', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } });
    const amountField = new Field('amount', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Float' } });

    serialNoField.default = { kind: 'DB_GENERATED', value: -1 };
    ownerNameField.default = { kind: 'DB_GENERATED', value: 'na' };
    amountField.default = { kind: 'DB_GENERATED', value: 101.101 };
    model.addField(serialNoField);
    model.addField(ownerNameField);
    model.addField(amountField);
    model.setPrimaryKey(['id']);

    dbschema.addModel(model);
    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('generates optional type but no default directive for fields with computed default values', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    const model = new Model('Account');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    const computedField = new Field('computed', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Float' } });

    computedField.default = { kind: 'DB_GENERATED', value: '(RAND() * RAND())' };
    model.addField(computedField);
    model.setPrimaryKey(['id']);

    dbschema.addModel(model);
    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('identifies the computed default values', () => {
    const testComputedExpressions = [
      'RAND()',
      'COS(PI())',
      'CONV(-17,10,-18)',
      'COS(CONV(-17,10,-18))',
      'LOG(CONV(-17,10,-18), 10)',
      '(RAND())',
      '(COS(PI()) * RAND())',
      '(CONV(-17,10,-18) + LOG(10, 100))',
      '(COS(CONV(-17,10,-18)))',
      '(LOG(CONV(-17,10,-18), 10))',
    ];

    testComputedExpressions.map((expr) => {
      expect(isComputeExpression(expr)).toEqual(true);
    });
  });

  it('test generate graphql schema on model with enum field', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    const model = new Model('Profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('type', { kind: 'Enum', name: 'Profile_type', values: ['Manager', 'Employee'] }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('generates primary key fields as required without the default directive added', () => {
    const dbschema = new Schema(new Engine('Postgres'));

    const model = new Model('Account');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } }));
    const serialNoField = new Field('serialNumber', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Int' } });
    const ownerNameField = new Field('ownerName', { kind: 'Scalar', name: 'String' });
    const amountField = new Field('amount', { kind: 'NonNull', type: { kind: 'Scalar', name: 'Float' } });

    model.addField(serialNoField);
    model.addField(ownerNameField);
    model.addField(amountField);
    model.setPrimaryKey(['id', 'serialNumber']);

    dbschema.addModel(model);
    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });
});
