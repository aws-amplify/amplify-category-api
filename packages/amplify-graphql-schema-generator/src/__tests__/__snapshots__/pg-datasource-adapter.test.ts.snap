// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testPostgresDataSourceAdapter exclude option should not import the given tables 1`] = `
"type Capital @model {
  id: Int! @primaryKey
  name: String
  countryId: Int @index(name: \\"countryId\\")
}

type Country @model {
  id: Int! @primaryKey
  name: String
}
"
`;

exports[`testPostgresDataSourceAdapter generate schema retains hasMany and belongsTo relationship 1`] = `
"type Blog @model {
  id: Int! @primaryKey
  name: String
  posts: [Post] @hasMany(references: [\\"userId\\"])
}

type Post @model {
  id: Int! @primaryKey
  content: String
  blogId: Int!
  blog: Blog @belongsTo(references: [\\"userId\\"])
}
"
`;

exports[`testPostgresDataSourceAdapter generate schema retains hasOne and belongsTo relationship and removes the non-relational fields added manually 1`] = `
"type Profile @model {
  id: Int! @primaryKey
  content: String
  userId: Int!
  user: User @belongsTo(references: [\\"userId\\"])
}

type User @model {
  id: Int! @primaryKey
  name: String
  profile: Profile @hasOne(references: [\\"userId\\"])
}
"
`;

exports[`testPostgresDataSourceAdapter generates a default directive and optional types for fields with literal default values 1`] = `
"type Account @model {
  id: Int! @primaryKey
  serialNumber: Int @default(value: \\"-1\\")
  ownerName: String @default(value: \\"na\\")
  amount: Float @default(value: \\"101.101\\")
}
"
`;

exports[`testPostgresDataSourceAdapter generates optional type but no default directive for fields with computed default values 1`] = `
"type Account @model {
  id: Int! @primaryKey
  computed: Float
}
"
`;

exports[`testPostgresDataSourceAdapter generates primary key fields as required without the default directive added 1`] = `
"type Account @model {
  id: Int! @primaryKey(sortKeyFields: [\\"serialNumber\\"])
  serialNumber: Int!
  ownerName: String
  amount: Float!
}
"
`;

exports[`testPostgresDataSourceAdapter include option should import only the given tables 1`] = `
"type Task @refersTo(name: \\"Tasks\\") @model {
  id: String! @primaryKey(sortKeyFields: [\\"title\\"])
  title: String @index(name: \\"tasks_title\\") @index(name: \\"tasks_title_description\\", sortKeyFields: [\\"description\\"])
  description: String
  priority: String
}
"
`;

exports[`testPostgresDataSourceAdapter test generate graphql schema from internal reprensentation 1`] = `
"type Capital @model {
  id: Int! @primaryKey
  name: String
  countryId: Int @index(name: \\"countryId\\")
}

type Country @model {
  id: Int! @primaryKey
  name: String
}

type Task @refersTo(name: \\"Tasks\\") @model {
  id: String! @primaryKey(sortKeyFields: [\\"title\\"])
  title: String @index(name: \\"tasks_title\\") @index(name: \\"tasks_title_description\\", sortKeyFields: [\\"description\\"])
  description: String
  priority: String
}
"
`;

exports[`testPostgresDataSourceAdapter test generate graphql schema on model with enum field 1`] = `
"type Profile @model {
  id: Int! @primaryKey
  name: String
  type: Profile_type
}

enum Profile_type {
  Manager
  Employee
}
"
`;
